packages/accounts-password/password_server.js

```diff
@@ -573,119 +573,185 @@ Accounts.sendResetPasswordEmail = function (userId, email) {
-
-  // Make sure the user exists, and email is in their addresses.
+Accounts.generateVerificationTokenRecord = function (userId, email, extraTokenData) {
+  // Make sure the user exists, and email is one of their addresses.
   var user = Meteor.users.findOne(userId);
   if (!user) {
-    throw new Error("Can't find user");
+    handleError("Can't find user");
   }
-  // pick the first email if we weren't passed an email.
-  if (!email && user.emails && user.emails[0]) {
-    email = user.emails[0].address;
+
+  // pick the first unverified email if we weren't passed an email.
+  if (!email) {
+    var emailRecord = _.find(user.emails || [], function (e) { return !e.verified; });
+    email = (emailRecord || {}).address;
+
+    if (!email) {
+      handleError("That user has no unverified email addresses.");
+    }
   }
+
   // make sure we have a valid email
   if (!email || !_.contains(_.pluck(user.emails || [], 'address'), email)) {
-    throw new Error("No such email for user.");
+    handleError("No such email for user.");
   }
 
-  var token = Random.secret();
-  var when = new Date();
   var tokenRecord = {
-    token: token,
-    email: email,
-    when: when,
-    reason: 'enroll'
+    token: Random.secret(),
+    // TODO: This should probably be renamed to "email" to match reset token record.
+    address: email,
+    when: new Date()
   };
-  Meteor.users.update(userId, {$set: {
-    "services.password.reset": tokenRecord
+
+  if (extraTokenData) {
+    _.extend(tokenRecord, extraTokenData);
+  }
+
+  return {email, user, tokenRecord};
+};
+
+/**
+ * @summary Saves the reset token record into database in user's document.
+ * @param {Object} user User for which to save the reset token record for.
+ * @param {Object} tokenRecord The reset token record to be saved.
+ * @importFromPackage accounts-base
+ */
+Accounts.saveResetTokenRecord = function (user, tokenRecord) {
+  Meteor.users.update({_id: user._id}, {$set: {
+    'services.password.reset': tokenRecord
   }});
 
   // before passing to template, update user object with new token
   Meteor._ensure(user, 'services', 'password').reset = tokenRecord;
+};
 
-  var enrollAccountUrl = Accounts.urls.enrollAccount(token);
+/**
+ * @summary Saves the e-mail verification token record into database in user's document.
+ * @param {Object} user User for which to save the e-mail verification token record for.
+ * @param {Object} tokenRecord The e-mail verification token record to be saved.
+ * @importFromPackage accounts-base
+ */
+Accounts.saveVerificationTokenRecord = function (user, tokenRecord) {
+  Meteor.users.update({_id: user._id}, {$push: {
+    'services.email.verificationTokens': tokenRecord
+  }});
 
+  // before passing to template, update user object with new token
+  Meteor._ensure(user, 'services', 'email');
+  if (!user.services.email.verificationTokens) {
+    user.services.email.verificationTokens = [];
+  }
+  user.services.email.verificationTokens.push(tokenRecord);
```

Comment:
Your re-implementation of this is still basically doing the same thing.  It'd be better if the calling-method just re-polled the database.  Again though, we could put this in the `generate*` method, avoid the public methods and in-line the update and it would be super clear.  Making a separate function for an update doesn't seem like it is necessary, especially if it's going to change/return a different object.